app/gui2/src/components/ComponentBrowser/__tests__/placement.test.ts

```diff
@@ -0,0 +1,653 @@
+          new Rect(new Vec2(1050, 690), defaultNodeSize),
+        ],
+        -1,
+      ),
+    ),
+    // 1791 + 100 + 24
+  ).toEqual<Placement>({ position: new Vec2(1915, 734) })
+  // Multiple nodes with smallest gap
+  expect(
+    previousNodeDictatedPlacement(
+      nodeSize,
+      environment2(
+        [
+          ...chain(
+            map(range(1000, 1401, 100), (x) => new Rect(new Vec2(x, 734), defaultNodeSize)),
+            map(range(1648, 1949, 100), (x) => new Rect(new Vec2(x, 734), defaultNodeSize)),
+          ),
+          new Rect(new Vec2(1050, 690), defaultNodeSize),
+        ],
+        -1,
+      ),
+    ),
+    // The smallest gap between the top of adjacent nodes this can fit into, is
+    // 100px width * 2 nodes + 24px gap * 2.
+  ).toEqual<Placement>({ position: new Vec2(1524, 734) })
+  // Multiple nodes with smallest gap (reverse)
+  expect(
+    previousNodeDictatedPlacement(
+      nodeSize,
+      environment2(
+        [
+          ...chain(
+            map(range(1948, 1647, -100), (x) => new Rect(new Vec2(x, 734), defaultNodeSize)),
+            map(range(1400, 999, -100), (x) => new Rect(new Vec2(x, 734), defaultNodeSize)),
+          ),
+          new Rect(new Vec2(1050, 690), defaultNodeSize),
+        ],
+        -1,
+      ),
+    ),
+  ).toEqual<Placement>({ position: new Vec2(1524, 734) })
+  // Multiple nodes with gap that is too small
+  expect(
+    previousNodeDictatedPlacement(
+      nodeSize,
+      environment2(
+        [
+          ...chain(
+            map(range(1000, 1401, 100), (x) => new Rect(new Vec2(x, 734), defaultNodeSize)),
+            map(range(1647, 1948, 100), (x) => new Rect(new Vec2(x, 734), defaultNodeSize)),
+          ),
+          new Rect(new Vec2(1050, 690), defaultNodeSize),
+        ],
+        -1,
+      ),
+    ),
+    // This gap is 1px smaller than the previous test - so, 1px too small.
+  ).toEqual<Placement>({ position: new Vec2(2071, 734), pan: new Vec2(1021, 44) })
+  // Multiple nodes with gap that is too small (each range reversed)
+  expect(
+    previousNodeDictatedPlacement(
+      nodeSize,
+      environment2(
+        [
+          ...chain(
+            map(range(1400, 999, -100), (x) => new Rect(new Vec2(x, 734), defaultNodeSize)),
+            map(range(1947, 1646, -100), (x) => new Rect(new Vec2(x, 734), defaultNodeSize)),
+          ),
+          new Rect(new Vec2(1050, 690), defaultNodeSize),
+        ],
+        -1,
+      ),
+    ),
+  ).toEqual<Placement>({ position: new Vec2(2071, 734), pan: new Vec2(1021, 44) })
+  // If fuzzing fails, then a new test case should be added above.
+  for (let i = 0; i < 100; i += 1) {
+    const nodes = Array.from(
+      { length: 20 },
+      () =>
+        new Rect(
+          new Vec2(Math.floor(Math.random() * 1000), Math.floor(Math.random() * 1000)),
+          new Vec2(Math.floor(Math.random() * 1000), Math.floor(Math.random() * 1000)),
+        ),
+    )
```

Comment:
This should be a property based test. We are using [fast-check vitest ingegration](https://www.npmjs.com/package/@fast-check/vitest) for that, so check it out. You can also use existing tests as a reference: 
https://github.com/enso-org/enso/blob/08c9ecb54037396a6ea26247c37b2bccb5ac1b4f/app/gui2/ydoc-server/__tests__/edits.test.ts#L86-L103